### PR TITLE
Targeted refresh worker

### PR DESCRIPTION
**Pull request by slemrmartin**
Targeted refresh of a job/workflow job is similar to order operation - it starts listening on Kafka queue.

- It can handle refresh of multiple jobs/workflows with the same Source.
- It requests Jobs, in case of Workflow its workflow nodes and jobs associated with the nodes.
Doesn't collect workflow nodes of child workflows (as we discussed)
- It also doesn't collect credentials for now
- All other relations are collected by full refresh so it's not needed to collect, refresh is creating `lazy_links` which are assigned to appropriate record in persister
- ~~removes** collecting of jobs/workflow job nodes from full refresh~~
- Refreshes are scheduled by https://github.com/slemrmartin/topological_inventory-scheduler/blob/master/lib/topological_inventory/scheduler/worker.rb (new service)

---

**Depends on**
- [x] Kafka topic `platform.topological-inventory.collector-ansible-tower` needs to be registered on CI/QA/Prod*
- [x] https://github.com/RedHatInsights/topological_inventory-ansible_tower/pull/117

Kafka Topic added to Prod [here](https://gitlab.cee.redhat.com/service/app-interface/-/merge_requests/8169)

---

**Dependents**
- https://github.com/slemrmartin/topological_inventory-scheduler
- https://github.com/RedHatInsights/e2e-deploy/pull/1868
---

[TPINVTRY-678](https://projects.engineering.redhat.com/browse/TPINVTRY-678)

### CODE REVIEW

**Comment by slemrmartin** [comment]
@syncrou I realized that targeted refresh is only new "operation". 
Question is, do we want to separate it as a new service(pod) (I made a new `bin/` script) or do we process it by operations worker?

EDIT: I would rather start it with new service and later join with full refresh service (see comment below)

**Comment by slemrmartin** [comment]
This is a preview, but I want to know your opinions, @lindgrenj6, @syncrou 

Refresh can be ordered by new service (
https://github.com/slemrmartin/topological_inventory-scheduler/blob/refresh-by-one-job/lib/topological_inventory/scheduler/worker.rb)
(..later it can also schedule full and incremental refreshes so we can remove dependency on orchestrator and scale collectors in a standard way. )

The most important part of this PR is https://github.com/RedHatInsights/topological_inventory-ansible_tower/pull/99/files#diff-fe657fb90a0328ada5e347f8eb7e1f97. It's able to process one job by one query at this moment.

**Comment by slemrmartin** [comment]
Update of ^^^:

Last commit allows update of multiple ServiceInstances (Jobs) with the same Source
Input is based on scheduler's master branch: https://github.com/slemrmartin/topological_inventory-scheduler/blob/master/lib/topological_inventory/scheduler/worker.rb

**Comment by lindgrenj6** [comment]
So after finally giving this a look - I'd like to propose having the "scheduler" part be a k8s `CronJob` like the availability checks, mostly because it's just sitting there in a while loop sleeping and then doing one direct task: putting a message on a queue https://github.com/slemrmartin/topological_inventory-scheduler/blob/ce2ded46e6c57229c060a57ae7e6a7c619e72b5c/lib/topological_inventory/scheduler/worker.rb#L31

I think if we just had this as a cronjob that ran every minute it would be easier on resources (since we wouldn't have a pod just sitting there sleeping most of the time), and it would be easier to configure since we can just change the Cron schedule on the fly.

**Comment by slemrmartin** [comment]
@lindgrenj6 I agree, I'm only afraid about the maintenance phase, you don't have `rsh` so you can't check the actual version of source code, the openshift UI seems to be much worse for jobs than pods and the openshift scheduler can be confused by some unknown problems (if you remember the problem with availability check cronjobs)

**Comment by lindgrenj6** [comment]
@slemrmartin My only counters there are:

1) The sources availability checkers have been working fine for the last few months since we got the configuration ironed out with `startingDeadlineSeconds`, `concurrencyPolicy` etc, they're pretty solid.

2) If we really did need to rsh into a pod to troubleshoot stuff on production its relatively simple to just create a `Pod` resource from the `jobTemplate` section of the cronjob, from there we can make the entrypoint `tail -f /dev/null` then `rsh` in and poke around. 

Here is an example of a pod from my db backup/restore scripts for catalog/approval OCP prod: https://github.com/lindgrenj6/insights-rds-db-dump-container/blob/master/catalog/resources/catalog_dumper.yml 
We could run the same image SHA256 that the `CronJob` would be using in this case.

File: lib/topological_inventory/ansible_tower/collector/service_catalog.rb
Comment (lindgrenj6) [inline lib/topological_inventory/ansible_tower/collector/service_catalog.rb]: Probably need to remove this.

File: lib/topological_inventory/ansible_tower/collector/service_catalog.rb
Comment (lindgrenj6) [inline lib/topological_inventory/ansible_tower/collector/service_catalog.rb]: For this parameter I think it could be a little clearer if we had the assignment in a block rather than having multiple post-modifiers, e.g:

```ruby
if refs.kind_of?(Array)
  opts[:id__in] = refs.join(',')
end
```

**Review by lindgrenj6 — COMMENTED** [review]
Just a few nits, I think I understand this well enough.

**Comment by slemrmartin** [comment]
I think it's ready. Targeted refresh has now only connection to cloud towers, the on-premise can be implemented in next PR.

File: spec/targeted_refresh/processor_spec.rb
Comment (gmcculloug) [inline spec/targeted_refresh/processor_spec.rb]: Rubocop pointed out these two lines for "Use hash rockets" which I would recommend resolving.  These lines are using mixed syntax which we should avoid.

**Comment by syncrou** [comment]
PR to add the new Kafka topic to Prod / Stage https://github.com/RedHatInsights/platform-mq/pull/80

File: lib/topological_inventory/ansible_tower/collector/service_catalog.rb
Comment (syncrou) [inline lib/topological_inventory/ansible_tower/collector/service_catalog.rb:135]: @slemrmartin My understanding was `if slug` that was a guarantee we were `on_premise`.

File: lib/topological_inventory/ansible_tower/messaging_client.rb
Comment (syncrou) [inline lib/topological_inventory/ansible_tower/messaging_client.rb:6]: should these constants be environment variables?

File: lib/topological_inventory/ansible_tower/messaging_client.rb
Comment (syncrou) [inline lib/topological_inventory/ansible_tower/messaging_client.rb:45]: Is this going to be the hostname like we're using in the other collectors?

File: lib/topological_inventory/ansible_tower/messaging_client.rb
Comment (syncrou) [inline lib/topological_inventory/ansible_tower/messaging_client.rb:52]: Same about the hostname as above.

File: lib/topological_inventory/ansible_tower/targeted_refresh/service_instance.rb
Comment (syncrou) [inline lib/topological_inventory/ansible_tower/targeted_refresh/service_instance.rb:36]: Is this tied to a paging limit of some type? Not sure why the number 20 is the default.

File: lib/topological_inventory/ansible_tower/targeted_refresh/service_instance.rb
Comment (syncrou) [inline lib/topological_inventory/ansible_tower/targeted_refresh/service_instance.rb:68]: Is the `if (is_missing = send(attr).blank?)` pattern used for readability?  I'm not sure if  `if(send(attr).blank?)` would be harder to read since we're iterating `required_params`.

File: lib/topological_inventory/ansible_tower/targeted_refresh/service_instance.rb
Comment (syncrou) [inline lib/topological_inventory/ansible_tower/targeted_refresh/service_instance.rb:93]: is there a reason `refresh_part_on_premise` is not public vs `refresh_part_cloud` is?

File: lib/topological_inventory/ansible_tower/targeted_refresh/service_instance.rb
Comment (syncrou) [inline lib/topological_inventory/ansible_tower/targeted_refresh/service_instance.rb:174]: Great idea!

File: lib/topological_inventory/ansible_tower/targeted_refresh/worker.rb
Comment (syncrou) [inline lib/topological_inventory/ansible_tower/targeted_refresh/worker.rb:57]: Is this going to have to change to default `ack` so it is inline with the receptor client issues we were seeing?

File: lib/topological_inventory/ansible_tower/targeted_refresh/service_instance.rb
Comment (lindgrenj6) [inline lib/topological_inventory/ansible_tower/targeted_refresh/service_instance.rb]: I think this method should be moved up (removing the public specifier) so we can just stick to one `private` definition rather than switching back and forth.

File: lib/topological_inventory/ansible_tower/targeted_refresh/service_instance.rb
Comment (lindgrenj6) [inline lib/topological_inventory/ansible_tower/targeted_refresh/service_instance.rb]: An authentication won't be present on receptor-enabled towers, at least not currently. So it shouldn't raise this unless we're not `on_premise?`

File: lib/topological_inventory/ansible_tower/targeted_refresh/service_instance.rb
Comment (lindgrenj6) [inline lib/topological_inventory/ansible_tower/targeted_refresh/service_instance.rb:44]: Can we get some metrics around this? Would be really nice to alert when `ServiceInstance#refresh` is failing a lot.

File: lib/topological_inventory/ansible_tower/targeted_refresh/service_instance.rb
Comment (lindgrenj6) [inline lib/topological_inventory/ansible_tower/targeted_refresh/service_instance.rb:150]: Yeah I think we could make use of the `Source` class in providers-common here, if we made the endpoint/auth methods public, would be nice to store the state there, definitely a follow-up though when we refactor a lot of the availability checking stuff like in #112.

**Review by lindgrenj6 — COMMENTED** [review]
Getting close on this one!

File: lib/topological_inventory/ansible_tower/targeted_refresh/service_instance.rb
Comment (slemrmartin) [reply lib/topological_inventory/ansible_tower/targeted_refresh/service_instance.rb]: @lindgrenj6 targeted refresh is for both cloud and on-premise

File: lib/topological_inventory/ansible_tower/targeted_refresh/service_instance.rb
Comment (slemrmartin) [reply lib/topological_inventory/ansible_tower/targeted_refresh/service_instance.rb:44]: I guess I added some TODO, I think this PR is big enough, so I'll rather add metrics in next PR

File: lib/topological_inventory/ansible_tower/collector/service_catalog.rb
Comment (slemrmartin) [reply lib/topological_inventory/ansible_tower/collector/service_catalog.rb:135]: @syncrou not always, if you want to use "find_all_by_url" (it's also ansible-tower-client's method), you need to provide slug

File: lib/topological_inventory/ansible_tower/messaging_client.rb
Comment (slemrmartin) [reply lib/topological_inventory/ansible_tower/messaging_client.rb:6]: this is a refactoring to have all topic's data together. I don't think we want to expose everything as an env variable. 
Moreover auto-creating topic is disabled, it doesn't make sense to me to have it configurable

File: lib/topological_inventory/ansible_tower/messaging_client.rb
Comment (slemrmartin) [reply lib/topological_inventory/ansible_tower/messaging_client.rb:45]: this is not an opts for receptor! This is constant consumer group shared accross ansible-tower-operations replicas

File: lib/topological_inventory/ansible_tower/messaging_client.rb
Comment (slemrmartin) [reply lib/topological_inventory/ansible_tower/messaging_client.rb:52]: the same answer as above :) that's not receptor's listener

File: lib/topological_inventory/ansible_tower/targeted_refresh/service_instance.rb
Comment (slemrmartin) [reply lib/topological_inventory/ansible_tower/targeted_refresh/service_instance.rb:36]: @syncrou I just specified something. The only limitation is max length of URL (which can't be computed in on-premise mode).
I guess 100 could be also fine. What do you suggest?

File: lib/topological_inventory/ansible_tower/targeted_refresh/service_instance.rb
Comment (slemrmartin) [reply lib/topological_inventory/ansible_tower/targeted_refresh/service_instance.rb:93]: both of them are private. Hm, obviously it's not a good idea to include public block inside private methods, I'll move it

File: lib/topological_inventory/ansible_tower/targeted_refresh/service_instance.rb
Comment (slemrmartin) [reply lib/topological_inventory/ansible_tower/targeted_refresh/service_instance.rb]: I wanted to group related methods together, but it seems it's not a good idea

File: lib/topological_inventory/ansible_tower/targeted_refresh/worker.rb
Comment (slemrmartin) [reply lib/topological_inventory/ansible_tower/targeted_refresh/worker.rb:57]: This is aligned with operations listener style. As the operation can take longer time, you can't have auto-ack because it can auto-timeout then. 
The receptor client issues were related to listener without `persist_ref`

File: lib/topological_inventory/ansible_tower/targeted_refresh/service_instance.rb
Comment (slemrmartin) [reply lib/topological_inventory/ansible_tower/targeted_refresh/service_instance.rb:68]: @syncrou this method is duplicated accross operations, so it'll be moved to providers-common in next refactoring. 
But I'm not sure what you are suggesting

File: lib/topological_inventory/ansible_tower/targeted_refresh/service_instance.rb
Comment (slemrmartin) [reply lib/topological_inventory/ansible_tower/targeted_refresh/service_instance.rb:174]: it'll wait to next PR because there are more duplicates accross operations

File: lib/topological_inventory/ansible_tower/targeted_refresh/service_instance.rb
Comment (lindgrenj6) [reply lib/topological_inventory/ansible_tower/targeted_refresh/service_instance.rb]: Yes, we just need to be checking in the `fetch_authentication` method if we're on premise and return a string or something (I just return a string constant in common) that way it doesn't raise the exception for receptor towers.

File: lib/topological_inventory/ansible_tower/targeted_refresh/service_instance.rb
Comment (syncrou) [reply lib/topological_inventory/ansible_tower/targeted_refresh/service_instance.rb:36]: 20 is probably fine. I was more interested in now that came to be. Seems like it can't be computed so any number as a default is okay.

File: lib/topological_inventory/ansible_tower/targeted_refresh/service_instance.rb
Comment (syncrou) [reply lib/topological_inventory/ansible_tower/targeted_refresh/service_instance.rb:68]: I don't see the need for the extra boolean:
```
          required_params.each do |attr|
            if (send(attr).blank?)
```

Probably worth a discussion later. Nothing to hold up this PR.

**Review by syncrou — COMMENTED** [review]
This looks good so far. A few items we can update in followup PRS.

There are a couple concerns @lindgrenj6 had yet, so I will wait till those are fully cleared up to approve.

File: lib/topological_inventory/ansible_tower/targeted_refresh/service_instance.rb
Comment (slemrmartin) [reply lib/topological_inventory/ansible_tower/targeted_refresh/service_instance.rb]: fixed

**Comment by slemrmartin** [comment]
@syncrou there is one possible issue: If this PR removes collecting of all service_instances, then ordering won't be completed until the cron job in e2e_deploy will be merged too. 

Especially I'm not sure how it is synchronized between standard code and e2e_deploy templates => app_interface

EDIT: Regarding that, won't it be more safe to remove the full collecting by following PR?

**Review by lindgrenj6 — APPROVED** [review]

**Pull request merged**
